### ISSUE TITLE
Replication of cache data to other nodes

### ISSUE THREAD

**Issue by shannonantony** [jiangwenyuan/nuster#35]
Implement replicating nuster node cache to other nodes

**Comment by jiangwenyuan** [comment]
That would be great, in fact, it's in the todo list. I'll work on this after HAProxy v1.9 migration and disk persistence implementation.  The progress is slow as I don't have much spare time as I used to.

**Comment by shannonantony** [comment]
Thanks for your quick response. 

Will you be using peers for the data syncing

**Comment by shannonantony** [comment]
Please let me know if you need any testing help. I could be the first
tester. thanks Jaison

**Comment by jiangwenyuan** [comment]
Hi, haven't looked into peers yet, that would be great if I can just use the sync feature from peers.

**Comment by igorescobar** [comment]
Meanwhile, what I'm doing to "solve" this issue is by using AWS EFS which allows me to have several nodes/instances/containers pointing to the same data volume for caching. So regardless of how many containers I have going up or down they are always sharing the same volume for caching which is great. 👍

**Comment by HugoS99** [comment]
I am trying to deploy nuster in a "cluster" (multiple containers) configuration, and the only problem I am having is that while they share the cache in a shared disk and not in memory, the disk loader doesn't synchronize the cache cross the different containers. Is there a way that I can force the disk loader to constantly ensure all containers have the updated cache?

**Comment by jiangwenyuan** [comment]
@HugoS99 What's your setup? a local disk directory mounted in multiple containers?

**Comment by HugoS99** [comment]
@jiangwenyuan It's a Kubernetes Cluster with a shared NFS folder across all containers and it resolves to a single shared folder and multiple containers accessing it

**Comment by jiangwenyuan** [comment]
@HugoS99 Should work then.  disk loader only loads the cache files metainfo onto memory on startup, it does not sync nor update cache files. Since you are using shared NFS folder, then all containers are seeing the same content, so there's no need to sync. If container 1 create a cache file in the shared NFS folder, container 2 can use that file to serve the identical request.
Make sure you are using `memory off disk on`

**Comment by HugoS99** [comment]
Yeah I know but what happens is since the load process only occurs on startup and is not checking constantly for new cached data what happens is that if a pod dies all other living pods don't "know" all the data that that one pod cached and as such they perform the request again

**Comment by HugoS99** [comment]
@jiangwenyuan Want me to create a pull request with my ugly fix?

**Comment by jiangwenyuan** [comment]
@HugoS99 You are right, and I now understand why #82 high cache MISS issue happens in multiple containers setup.

When the loader is done, nuster does not check disk anymore and only relies on memory, so if another container creates a new file, it does not know.

Thanks for the proposal, but constantly loading does not solve this problem(because it may happen before next round loading) and is not necessary in single nuster setup.

I probably will add a new mode for multiple nuster setup.

**Comment by HugoS99** [comment]
I agree that a new "cluster" mode should be created, but I believe the constant read is the most generic and more natural way to facilitate this feature the best solution would be to allow the user to set the time between round loading you may see my implementation for reference (where each cycle takes 300ms) simply allow a user to set that value with the clarification that there may be missed caches with this setup, but a user can simply put 0 to ensure the best possible synchronization ("but there still may be missed cache")

**Comment by jiangwenyuan** [comment]
nst_disk_load does not load all files but several files. If you have like 10M files, it will cause much time to complete the load. And it will generate lots IO.

**Comment by HugoS99** [comment]
So  a method that compared all the files already loaded with the not loaded is required with that the solution would be valid right?

**Comment by jiangwenyuan** [comment]
Yeah. Currently a simplest solution pops up into my head is a new mode that always check disk if does not exists in memory.

So currently the logic is (for disk on memory off):

* check memory(does disk.file exist?)
   * yes, use disk.file
    * no,  loaded?
       * no, check disk (does x/xx/uuid file exist?)
       * yes, goto backend

So the new mode is:

* check memory(does disk.file exist?)
   * yes, use disk.file
    * no,  loaded?
       * no, check disk (does x/xx/uuid file exist?)
       * yes and (new mode), check disk (does x/xx/uuid file exist?)
       * else, goto backend

**Comment by HugoS99** [comment]
That looks like a great simple solution! If you tell the file where that logic is performed I can try and implement an initial draft solution!

**Comment by jiangwenyuan** [comment]
Cool! Thanks a lot! I'll update this comment later.

**Comment by HugoS99** [comment]
Hey @jiangwenyuan just finished the initial implementation the only thing missing is the delete portion I didn't understand the code so I thought I should ask you do you want a PR for this?

**Comment by jiangwenyuan** [comment]
@HugoS99 Sure, thanks
